### DISCUSSION TITLE
Unexpected Inverter data

### DISCUSSION THREAD

**Discussion by kkrankall** [CharlesGodwin/pymagnum#42] [General]
First off everything works great. It's just that sometimes I get the strange data about an inverter that is not mine. Not sure if it's because my packet count is low (25 packets). I set it low so there is less latency as I'm calling the API every 5 seconds. I have 2 MS4024PAEs but every now in then I pick up some other models (MS2012E, MS4448PAE, MS4124E) that would then cause some crazy voltage readings

<img width="1365" alt="Screen Shot 2022-10-27 at 9 47 10 AM" src="https://user-images.githubusercontent.com/15572661/198302135-95d77be2-cd10-4ac8-94d2-a5e0898c7af3.png">

<img width="873" alt="image" src="https://user-images.githubusercontent.com/15572661/198305608-a84e73a0-6d92-4915-aa0f-105edd49da50.png">

In HA (Home Assistant) I have added logic to setting the sensor state value to make sure It's only recording data for the correct Inverter Model (MS4024PAE) and that if the API is down for some reason it will just keep the last value so I dont have a bunch of weird looking data. 

`
template:
  - sensor:
      - name: "Magnum Inverter Volts"
        unit_of_measurement: "V"
        state: >
            {% set model = state_attr('sensor.magnum_inverter_master', 'model_text')  %}
            {% set ignore = ['unknown', 'unavailable'] %}
            {% if state_attr('sensor.magnum_inverter_master', 'vdc') not in ignore %}
            {% if model == 'MS4024PAE' %}
            {{ state_attr('sensor.magnum_inverter_master', 'vdc') }}
            {% else %}
            {{ states.sensor.magnum_inverter_volts.state | float(default=0)}}
            {% endif %}
            {% else %}
            {{ states.sensor.magnum_inverter_volts.state | float(default=0)}}
            {% endif %}
        device_class: voltage
        state_class: measurement
`

Just wondering if anyone else is getting strange data

Thank you,

Kevin

**Comment by CharlesGodwin** [comment]
You're the first to report it

Charles
________________________________
From: kkrankall ***@***.***>
Sent: Thursday, October 27, 2022 10:03:39 AM
To: CharlesGodwin/pymagnum ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [CharlesGodwin/pymagnum] Unexpected Inverter data (Discussion #42)


First off everything works great. It's just that sometimes I get the strange data about an inverter that is not mine. Not sure if it's because my packet count is low (25 packets). I set it low so there is less latency as I'm calling the API every 5 seconds. I have 2 MS4024PAEs but every now in then I pick up some other models (MS2012E, MS4448PAE, MS4124E) that would then cause some crazy voltage readings

[Screen Shot 2022-10-27 at 9 47 10 AM]<https://user-images.githubusercontent.com/15572661/198302135-95d77be2-cd10-4ac8-94d2-a5e0898c7af3.png>

[image]<https://user-images.githubusercontent.com/15572661/198305608-a84e73a0-6d92-4915-aa0f-105edd49da50.png>

In HA (Home Assistant) I have added logic to setting the sensor state value to make sure It's only recording data for the correct Inverter Model (MS4024PAE) and that if the API is down for some reason it will just keep the last value so I dont have a bunch of weird looking data.

`
template:

  *   sensor:
     *   name: "Magnum Inverter Volts"
unit_of_measurement: "V"
state: >
{% set model = state_attr('sensor.magnum_inverter_master', 'model_text') %}
{% set ignore = ['unknown', 'unavailable'] %}
{% if state_attr('sensor.magnum_inverter_master', 'vdc') not in ignore %}
{% if model == 'MS4024PAE' %}
{{ state_attr('sensor.magnum_inverter_master', 'vdc') }}
{% else %}
{{ states.sensor.magnum_inverter_volts.state | float(default=0)}}
{% endif %}
{% else %}
{{ states.sensor.magnum_inverter_volts.state | float(default=0)}}
{% endif %}
device_class: voltage
state_class: measurement
`

Just wondering if anyone else is getting strange data

Thank you,

Kevin

**Comment by kkrankall** [comment]
Also I sometime get a PT100 charge controller (Had one for a week before it gave up the ghost) and it will show up with NaN in a few values. Home Assistant doesn't like this and says "REST result could not be parsed as JSON"

Here is what the JSON data looks like

```
[
    {
        "comm_device": "/dev/ttyUSB1",
        "data": {
            "datetime": "2022-11-06T13:44:22-05:00",
            "devices": [
                {
                    "device": "INVERTER",
                    "data": {
                        "revision": "6.1",
                        "mode": 64,
                        "mode_text": "INVERT",
                        "fault": 0,
                        "fault_text": "None",
                        "vdc": 26.2,
                        "adc": 16.0,
                        "VACout": 116.0,
                        "VACin": 180.0,
                        "invled": 1,
                        "invled_text": "On",
                        "chgled": 0,
                        "chgled_text": "Off",
                        "bat": 25.0,
                        "tfmr": 50.0,
                        "fet": 35.0,
                        "model": 107,
                        "model_text": "MS4024PAE",
                        "stackmode": 1,
                        "stackmode_text": "Parallel stack - master",
                        "AACin": 0.0,
                        "AACout": 3.0,
                        "Hz": 60.0
                    }
                },
                {
                    "device": "REMOTE",
                    "data": {
                        "revision": 4.0,
                        "searchwatts": 5,
                        "batterysize": 50,
                        "battype": 0,
                        "absorb": 24.0,
                        "chargeramps": 100,
                        "ainput": 5,
                        "parallel": 50,
                        "lbco": 10.5,
                        "vaccutout": 255.0,
                        "batteryefficiency": 0,
                        "vsfloat": 24.0,
                        "vEQ": 24.0,
                        "absorbtime": 2.0
                    }
                },
                {
                    "device": "BMK",
                    "data": {
                        "revision": "1.0",
                        "soc": 64,
                        "vdc": 26.14,
                        "adc": -4.4,
                        "vmin": 17.28,
                        "vmax": 29.98,
                        "amph": -143,
                        "amphtrip": 6553.5,
                        "amphout": 20000,
                        "Fault": 1,
                        "Fault_Text": "Normal"
                    }
                },
                {
                    "device": "RTR",
                    "data": {
                        "revision": "4"
                    }
                },
                {
                    "device": "PT100",
                    "data": {
                        "revision": "1.x",
                        "address": 0,
                        "mode": 0,
                        "mode_text": "",
                        "mode_hex": "",
                        "regulation": 0,
                        "regulation_text": "",
                        "fault": 0,
                        "fault_text": "",
                        "battery": 0.0,
                        "battery_amps": 0.0,
                        "pv_voltage": 0.0,
                        "charge_time": 0.0,
                        "target_battery_voltage": 0.0,
                        "relay_state": 1,
                        "alarm_state": 0,
                        "fan_on": 0,
                        "day": 0,
                        "battery_temperature": 0.0,
                        "inductor_temperature": 0.0,
                        "fet_temperature": 0.0,
                        "lifetime_kwhrs": NaN,
                        "resettable_kwhrs": NaN,
                        "ground_fault_current": NaN,
                        "nominal_battery_voltage": 12.0,
                        "stacker_info": 0,
                        "dip_switches": "00000000",
                        "model": "",
                        "output_current_rating": NaN,
                        "input_voltage_rating": NaN
                    },
                    "trace": []
                }
            ]
        }
    },
    {
        "comm_device": "/dev/ttyUSB0",
        "data": {
            "datetime": "2022-11-06T13:44:22-05:00",
            "devices": [
                {
                    "device": "INVERTER",
                    "data": {
                        "revision": "6.1",
                        "mode": 80,
                        "mode_text": "Inverter_Standby",
                        "fault": 0,
                        "fault_text": "None",
                        "vdc": 26.1,
                        "adc": 0.0,
                        "VACout": 16.0,
                        "VACin": 0.0,
                        "invled": 1,
                        "invled_text": "On",
                        "chgled": 0,
                        "chgled_text": "Off",
                        "bat": 25.0,
                        "tfmr": 30.0,
                        "fet": 28.0,
                        "model": 107,
                        "model_text": "MS4024PAE",
                        "stackmode": 2,
                        "stackmode_text": "Parallel stack - slave",
                        "AACin": 0.0,
                        "AACout": 0.0,
                        "Hz": 0.0
                    }
                },
                {
                    "device": "REMOTE",
                    "data": {
                        "revision": 25.5,
                        "searchwatts": 5,
                        "batterysize": 400,
                        "battype": 5,
                        "absorb": 0,
                        "chargeramps": 0,
                        "ainput": 105,
                        "parallel": 80,
                        "lbco": 2.0,
                        "vaccutout": 96.0,
                        "batteryefficiency": 0
                    }
                },
                {
                    "device": "BMK",
                    "data": {
                        "revision": "1.0",
                        "soc": 64,
                        "vdc": 26.14,
                        "adc": -4.4,
                        "vmin": 17.28,
                        "vmax": 29.98,
                        "amph": -143,
                        "amphtrip": 6553.5,
                        "amphout": 20000,
                        "Fault": 1,
                        "Fault_Text": "Normal"
                    }
                }
            ]
        }
    }
]
```

**Reply by CharlesGodwin** [reply]
NaN is a term to define that a numeric variable does not contain a valid numeric value (**N**ot **a** **N**umber). In this situation it means  that PT-100 has not provided any information.  These variables are provided as one packet from PT-100 and NaN implies no such packet has been received.
I was under the impression that NaN was a convention in JSON syntax but I guess not. The only thing I can do is delete all these variables from the JSON output if this packet has not been received. Alas, Magnum has not been helpful with providing any PT-100 data.

This will take a few days.

<pre>
"lifetime_kwhrs"
"resettable_kwhrs"
"ground_fault_current"
"nominal_battery_voltage"
"stacker_info"
'dip_switches'
"model"
"output_current_rating"
"input_voltage_rating"
</pre>

**Reply by CharlesGodwin** [reply]
I got lucky. Version 2.0.2 is available.

**Reply by kkrankall** [reply]
You're amazing. I will give it a try

**Reply by CharlesGodwin** [reply]
don't thank me until it works. ;)

**Reply by kkrankall** [reply]
It works! Thank you.

```
{
                    "device": "PT100",
                    "data": {
                        "revision": "1.x",
                        "address": 0,
                        "mode": 0,
                        "mode_text": "",
                        "mode_hex": "",
                        "regulation": 0,
                        "regulation_text": "",
                        "fault": 0,
                        "fault_text": "",
                        "battery": 0.0,
                        "battery_amps": 0.0,
                        "pv_voltage": 0.0,
                        "charge_time": 0.0,
                        "target_battery_voltage": 0.0,
                        "relay_state": 1,
                        "alarm_state": 0,
                        "fan_on": 0,
                        "day": 0,
                        "battery_temperature": 0.0,
                        "inductor_temperature": 0.0,
                        "fet_temperature": 0.0
                    },
                    "trace": []
                }
```

**Reply by CharlesGodwin** [reply]
thanks for the feedback.
